group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2930,72 +2930,86 @@ public CoordinatorResult<LeaveGroupResponseData, Record> genericGroupLeave(
-                                .setErrorCode(Errors.forException(e).code())
-                        );
-                    }
+                    memberResponses.add(
+                        new MemberResponse()
+                            .setMemberId(member.memberId())
+                            .setGroupInstanceId(member.groupInstanceId())
+                            .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code())
+                    );
+                }
+            } else if (group.isPendingMember(member.memberId())) {
+                coordinatorResults.add(removePendingMemberAndUpdateGenericGroup(group, member.memberId()));
+                timer.cancel(genericGroupHeartbeatKey(group.groupId(), member.memberId()));
+                log.info("Pending member {} has left group {} through explicit `LeaveGroup` request. Reason: {}",
+                    member.memberId(), group.groupId(), reason);
+
+                memberResponses.add(
+                    new MemberResponse()
+                        .setMemberId(member.memberId())
+                        .setGroupInstanceId(member.groupInstanceId())
+                );
+            } else {
+                try {
+                    group.validateMember(member.memberId(), member.groupInstanceId(), "leave-group");
+                    coordinatorResults.add(removeCurrentMemberFromGenericGroup(
+                        group,
+                        member.memberId(),
+                        reason
+                    ));
+                    memberResponses.add(
+                        new MemberResponse()
+                            .setMemberId(member.memberId())
+                            .setGroupInstanceId(member.groupInstanceId())
+                    );
+                } catch (KafkaException e) {
+                    memberResponses.add(
+                        new MemberResponse()
+                            .setMemberId(member.memberId())
+                            .setGroupInstanceId(member.groupInstanceId())
+                            .setErrorCode(Errors.forException(e).code())
+                    );
                 }
             }
-            return new CoordinatorResult<>(
-                coordinatorResult.records(),
-                new LeaveGroupResponseData()
-                    .setMembers(memberResponses),
-                coordinatorResult.appendFuture()
-            );
+        }
+
+        List<CoordinatorResult<Void, Record>> results = coordinatorResults.stream()
+            .filter(result -> result != EMPTY_RESULT)
+            .collect(Collectors.toList());
+
+        CoordinatorResult<Void, Record> coordinatorResult = EMPTY_RESULT;
+
+        if (results.size() > 1) {
+            throw new IllegalStateException("Expected max 1 non-empty result but found: " + results.size());
+        } else if (results.size() == 1) {
+            coordinatorResult = coordinatorResults.get(0);
+        }
```

Comment:
This seems a bit fragile to me. How about creating a new CoordinatorResult which contains all the records generated and a write future that will be used to complete all the write future of all the generated CoordinatorResults? I suppose that we could assume that a CoordinatorResult without records can be ignored. If we have only one CoordinatorResult, we could just return it as you did here. I think that we could wrap this in a new class, say CoordinatorResultAggregator/Builder. What do you think?

Reply:
I'm not a fan because essentially we only want 1 coordinator result per API call because there's a 1-1 mapping of request to CoordinatorResult#response. So, even if we wrap this around a new class we still need to condense it into a single result.

If we expect a maximum of 1 non-empty coordinator result, shouldn't we fail fast instead and fix the bug?

Reply:
As discussed offline, I have changed the code to perform a single check at the end to either rebalance/complete join/do nothing.

a caveat is that completing joins will only be done if it is not the initial rebalance (similar to our discussion here: https://github.com/apache/kafka/pull/14147#discussion_r1316322730) for pending member leaving.

A pending member leave and a valid member leave will both result in the same maybeCompleteJoinPhase().